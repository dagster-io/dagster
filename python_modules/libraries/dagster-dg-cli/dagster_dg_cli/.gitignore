docs/packages
docs/node_modules
