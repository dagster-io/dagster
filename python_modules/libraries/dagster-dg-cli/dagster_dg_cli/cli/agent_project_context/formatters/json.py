"""JSON formatter for project context."""

import json

from dagster_dg_cli.cli.agent_project_context.formatters import BaseFormatter
from dagster_dg_cli.cli.agent_project_context.models import ProjectContext


class JsonFormatter(BaseFormatter):
    """Formats project context as JSON."""

    def format(self, context: ProjectContext) -> str:
        """Format project context as JSON."""
        # Convert record objects to dictionaries for JSON serialization
        context_dict = {
            "working_directory": context.working_directory,
            "project_root": context.project_root,
            "project_name": context.project_name,
            "components": [
                {
                    "key": comp.key,
                    "summary": comp.summary,
                    "namespace": comp.namespace,
                    "name": comp.name,
                }
                for comp in context.components
            ],
            "integrations": [
                {
                    "name": integ.name,
                    "description": integ.description,
                    "pypi": integ.pypi,
                }
                for integ in context.integrations
            ],
            "project_structure": {
                "has_pyproject_toml": context.project_structure.has_pyproject_toml,
                "has_dbt_project": context.project_structure.has_dbt_project,
                "dbt_project_paths": context.project_structure.dbt_project_paths,
                "python_packages": context.project_structure.python_packages,
            },
            "dbt_schemas": {
                key: {
                    "description": schema.description,
                    "scaffold_params_schema": schema.scaffold_params_schema,
                    "component_schema": schema.component_schema,
                }
                for key, schema in context.dbt_schemas.items()
            },
        }
        return json.dumps(context_dict, indent=2)
