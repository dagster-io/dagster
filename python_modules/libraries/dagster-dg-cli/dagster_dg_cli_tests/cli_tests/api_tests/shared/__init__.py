"""Shared utilities for API test recordings."""
