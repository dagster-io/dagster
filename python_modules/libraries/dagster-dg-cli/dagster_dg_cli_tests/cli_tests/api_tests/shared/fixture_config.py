"""Configuration for API fixture recording scenarios."""

from dataclasses import dataclass
from typing import Optional


@dataclass
class FixtureScenario:
    """Configuration for a single fixture recording scenario."""

    command: str
    depends_on: Optional[list[str]] = None  # Dependencies (run IDs, etc.)
