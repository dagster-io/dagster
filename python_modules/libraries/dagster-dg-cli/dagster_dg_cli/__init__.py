from dagster_shared.libraries import DagsterLibraryRegistry

from dagster_dg_cli.version import __version__ as __version__

DagsterLibraryRegistry.register("dagster-dg-cli", __version__)
