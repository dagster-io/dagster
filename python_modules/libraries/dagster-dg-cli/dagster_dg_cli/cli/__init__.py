import os
from pathlib import Path

import click
from dagster_dg_core.config import normalize_cli_config
from dagster_dg_core.context import DgContext
from dagster_dg_core.shared_options import dg_global_options, dg_path_options
from dagster_dg_core.utils import DG_CLI_MAX_OUTPUT_WIDTH, DgClickGroup

from dagster_dg_cli.cli.check import check_group
from dagster_dg_cli.cli.dev import dev_command
from dagster_dg_cli.cli.docs import docs_group
from dagster_dg_cli.cli.launch import launch_command
from dagster_dg_cli.cli.list import list_group
from dagster_dg_cli.cli.mcp_server import mcp_group
from dagster_dg_cli.cli.plus import plus_group
from dagster_dg_cli.cli.scaffold import scaffold_group
from dagster_dg_cli.cli.utils import utils_group
from dagster_dg_cli.version import __version__


def create_dg_cli():
    @click.group(
        name="dg",
        commands={
            "check": check_group,
            "docs": docs_group,
            "utils": utils_group,
            "launch": launch_command,
            "list": list_group,
            "scaffold": scaffold_group,
            "dev": dev_command,
            "plus": plus_group,
            "mcp": mcp_group,
        },
        context_settings={
            "max_content_width": DG_CLI_MAX_OUTPUT_WIDTH,
            "help_option_names": ["-h", "--help"],
        },
        invoke_without_command=True,
        cls=DgClickGroup,
    )
    @dg_path_options
    @dg_global_options
    @click.option(
        "--clear-cache",
        is_flag=True,
        help="Clear the cache.",
        default=False,
    )
    @click.option(
        "--install-completion",
        is_flag=True,
        help="Automatically detect your shell and install a completion script for the `dg` command. This will append to your shell startup file.",
        default=False,
    )
    @click.version_option(__version__, "--version", "-v")
    def group(
        install_completion: bool,
        clear_cache: bool,
        path: Path,
        **global_options: object,
    ):
        """CLI for managing Dagster projects."""
        os.environ["DAGSTER_IS_DEV_CLI"] = "1"

        context = click.get_current_context()
        if install_completion:
            import dagster_dg_core.completion

            dagster_dg_core.completion.install_completion(context)
            context.exit(0)
        elif clear_cache:
            from dagster_dg_core.cache import DgCache

            cli_config = normalize_cli_config(global_options, context)
            dg_context = DgContext.from_file_discovery_and_command_line_config(path, cli_config)
            # Normally we would access the cache through the DgContext, but cache is currently
            # disabled outside of a project context. When that restriction is lifted, we will change
            # this to access the cache through the DgContext.
            cache = DgCache.from_config(dg_context.config)
            cache.clear_all()
            if context.invoked_subcommand is None:
                context.exit(0)

        elif context.invoked_subcommand is None:
            click.echo(context.get_help())
            context.exit(0)

    return group


ENV_PREFIX = "DAGSTER_DG"
cli = create_dg_cli()


def main():
    cli(auto_envvar_prefix=ENV_PREFIX)
