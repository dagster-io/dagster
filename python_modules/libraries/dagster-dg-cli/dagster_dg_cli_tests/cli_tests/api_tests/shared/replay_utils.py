"""Shared utilities for replaying recorded GraphQL responses in tests."""

import json
from collections.abc import Mapping
from pathlib import Path
from typing import Any, Optional

from dagster_dg_cli.utils.plus.gql_client import DagsterPlusGraphQLClient


class ReplayClient(DagsterPlusGraphQLClient):
    """GraphQL client that replays recorded responses."""

    def __init__(self, responses: list[dict[str, Any]]):
        # Don't call super().__init__ as we don't need the real GraphQL client
        self.responses = responses
        self.call_index = 0

    def execute(self, query: str, variables: Optional[Mapping[str, Any]] = None) -> dict:
        """Return next recorded response."""
        if self.call_index >= len(self.responses):
            raise ValueError(f"Exhausted {len(self.responses)} responses")
        response = self.responses[self.call_index]
        self.call_index += 1
        return response


def load_recorded_graphql_responses(
    recordings_dir: Path, scenario_name: str
) -> list[dict[str, Any]]:
    """Load GraphQL response recordings for a given scenario.

    Args:
        recordings_dir: Directory containing the recordings
        scenario_name: Name of the scenario to load

    Returns:
        List of recorded GraphQL responses
    """
    scenario_folder = recordings_dir / scenario_name

    if not scenario_folder.exists():
        raise ValueError(f"Recording scenario not found: {scenario_folder}")

    json_files = sorted([f for f in scenario_folder.glob("*.json") if f.name[0:2].isdigit()])

    if not json_files:
        raise ValueError(f"No numbered JSON files found in {scenario_folder}")

    responses = []
    for json_file in json_files:
        with open(json_file) as f:
            responses.append(json.load(f))

    return responses
