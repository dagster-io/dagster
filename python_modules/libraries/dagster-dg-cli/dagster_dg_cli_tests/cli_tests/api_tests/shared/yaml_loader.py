"""YAML loading utilities for fixture scenario configurations."""

from pathlib import Path

import yaml

from dagster_dg_cli_tests.cli_tests.api_tests.shared.fixture_config import FixtureScenario


def load_fixture_scenarios_from_yaml(yaml_file: Path) -> dict[str, FixtureScenario]:
    """Load fixture scenarios from a YAML file.

    Args:
        yaml_file: Path to YAML file containing fixture scenario definitions

    Returns:
        Dictionary mapping fixture names to FixtureScenario objects

    Raises:
        ValueError: If YAML file doesn't exist or has invalid format
    """
    if not yaml_file.exists():
        raise ValueError(f"YAML fixture file not found: {yaml_file}")

    try:
        with open(yaml_file) as f:
            yaml_data = yaml.safe_load(f) or {}
    except yaml.YAMLError as e:
        raise ValueError(f"Invalid YAML in {yaml_file}: {e}")

    fixture_scenarios = {}
    for fixture_name, config_data in yaml_data.items():
        if not isinstance(config_data, dict):
            raise ValueError(
                f"Invalid fixture config for '{fixture_name}': expected dict, got {type(config_data)}"
            )

        if "command" not in config_data:
            raise ValueError(f"Missing 'command' field for fixture '{fixture_name}'")

        fixture_scenarios[fixture_name] = FixtureScenario(
            command=config_data["command"], depends_on=config_data.get("depends_on")
        )

    return fixture_scenarios
