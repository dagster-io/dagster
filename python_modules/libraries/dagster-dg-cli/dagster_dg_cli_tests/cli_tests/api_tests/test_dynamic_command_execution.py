import json
from collections.abc import Iterator, Mapping
from pathlib import Path
from typing import Any, Optional

import pytest
from click.testing import CliRunner
from dagster_dg_cli.cli.api.client import DgApiTestContext
from dagster_dg_cli.utils.plus.gql_client import DagsterPlusGraphQLClient

from dagster_dg_cli_tests.cli_tests.api_tests.shared.yaml_loader import (
    load_fixture_scenarios_from_yaml,
)


class ReplayClient(DagsterPlusGraphQLClient):
    """GraphQL client that replays recorded responses."""

    def __init__(self, responses: list[dict[str, Any]]):
        # Don't call super().__init__ as we don't need the real GraphQL client
        self.responses = responses
        self.call_index = 0

    def execute(self, query: str, variables: Optional[Mapping[str, Any]] = None) -> dict:
        """Return next recorded response."""
        if self.call_index >= len(self.responses):
            raise ValueError(f"Exhausted {len(self.responses)} responses")
        response = self.responses[self.call_index]
        self.call_index += 1
        return response


def discover_scenario_recordings() -> Iterator[tuple[str, str, str]]:
    """Discover all scenario recordings across API test domains."""
    api_tests_dir = Path(__file__).parent

    domain_dirs = [d for d in api_tests_dir.iterdir() if d.is_dir() and d.name.endswith("_tests")]

    for domain_dir in domain_dirs:
        scenarios_yaml = domain_dir / "scenarios.yaml"
        if not scenarios_yaml.exists():
            continue

        domain = domain_dir.name.replace("_tests", "")
        recording_scenarios = load_fixture_scenarios_from_yaml(scenarios_yaml)

        for recording_name, recording_config in recording_scenarios.items():
            yield (domain, recording_name, recording_config.command)


def load_recorded_graphql_responses(domain: str, scenario_name: str) -> list[dict[str, Any]]:
    """Load GraphQL response recordings for a given domain and scenario."""
    scenario_folder = Path(__file__).parent / f"{domain}_tests" / "recordings" / scenario_name

    if not scenario_folder.exists():
        raise ValueError(f"Recording scenario not found: {scenario_folder}")

    json_files = sorted([f for f in scenario_folder.glob("*.json") if f.name[0:2].isdigit()])

    if not json_files:
        raise ValueError(f"No numbered JSON files found in {scenario_folder}")

    responses = []
    for json_file in json_files:
        with open(json_file) as f:
            responses.append(json.load(f))

    return responses


class TestDynamicCommandExecution:
    """Test all commands from YAML scenarios against recorded GraphQL responses."""

    @pytest.mark.parametrize("domain,scenario_name,command", list(discover_scenario_recordings()))
    def test_command_execution(self, domain: str, scenario_name: str, command: str, snapshot):
        """Test executing a command against its recorded GraphQL responses.

        Uses Click's dependency injection to provide a test client factory.
        No mocking required - just inject a context with our replay client.
        """
        from dagster_dg_cli.cli import cli as root_cli

        # Load GraphQL responses for this scenario
        graphql_responses = load_recorded_graphql_responses(domain, scenario_name)

        # Create replay client
        replay_client = ReplayClient(graphql_responses)

        # Create context with test factory
        test_context = DgApiTestContext(client_factory=lambda config: replay_client)

        # Run command with injected context - SINGLE INJECTION POINT
        runner = CliRunner()
        args = command.split()[1:]  # Skip 'dg'

        # Click's obj parameter passes our context through
        result = runner.invoke(root_cli, args, obj=test_context)

        # Snapshot testing
        output_type = "json" if "--json" in command else "text"
        snapshot_name = f"{domain}_{scenario_name}_{output_type}"

        if "--json" in command:
            try:
                parsed_output = json.loads(result.output)
                assert parsed_output == snapshot(name=snapshot_name)
            except json.JSONDecodeError:
                assert result.output == snapshot(name=snapshot_name)
        else:
            assert result.output == snapshot(name=snapshot_name)

        if "error" not in scenario_name.lower():
            assert result.exit_code == 0
