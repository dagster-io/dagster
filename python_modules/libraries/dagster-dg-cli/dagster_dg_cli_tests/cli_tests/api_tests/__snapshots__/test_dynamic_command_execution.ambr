# serializer version: 1
# name: TestDynamicCommandExecution.test_command_execution[asset_error_asset_not_found_json]
  '''
  {"error": "Asset not found: nonexistent-asset"}
  Error: Failed to get asset: Asset not found: nonexistent-asset
  
  '''
# ---
# name: TestDynamicCommandExecution.test_command_execution[asset_error_malformed_asset_key_json]
  '''
  {"error": "Asset not found: ''"}
  Error: Failed to get asset: Asset not found: ''
  
  '''
# ---
# name: TestDynamicCommandExecution.test_command_execution[asset_success_assets_with_cursor_json]
  dict({
    'cursor': None,
    'has_more': False,
    'items': list([
    ]),
  })
# ---
# name: TestDynamicCommandExecution.test_command_execution[asset_success_multiple_assets_json]
  dict({
    'cursor': '["aws", "cloud-prod", "workspace_staging_pipelines"]',
    'has_more': True,
    'items': list([
      dict({
        'asset_key': 'aws/cloud-prod/user_roles',
        'asset_key_parts': list([
          'aws',
          'cloud-prod',
          'user_roles',
        ]),
        'description': 'Snowflake stages for AWS data, creates new stages for new assets, refreses existing stages.',
        'group_name': 'aws_stages',
        'id': 'dagster_open_platform.__repository__.["aws", "cloud-prod", "user_roles"]',
        'kinds': list([
        ]),
        'metadata_entries': list([
        ]),
      }),
      dict({
        'asset_key': 'aws/cloud-prod/workspace_staging_asset_checks',
        'asset_key_parts': list([
          'aws',
          'cloud-prod',
          'workspace_staging_asset_checks',
        ]),
        'description': 'Snowflake stages for AWS data, creates new stages for new assets, refreses existing stages.',
        'group_name': 'aws_stages',
        'id': 'dagster_open_platform.__repository__.["aws", "cloud-prod", "workspace_staging_asset_checks"]',
        'kinds': list([
        ]),
        'metadata_entries': list([
        ]),
      }),
      dict({
        'asset_key': 'aws/cloud-prod/workspace_staging_assets',
        'asset_key_parts': list([
          'aws',
          'cloud-prod',
          'workspace_staging_assets',
        ]),
        'description': 'Snowflake stages for AWS data, creates new stages for new assets, refreses existing stages.',
        'group_name': 'aws_stages',
        'id': 'dagster_open_platform.__repository__.["aws", "cloud-prod", "workspace_staging_assets"]',
        'kinds': list([
        ]),
        'metadata_entries': list([
        ]),
      }),
      dict({
        'asset_key': 'aws/cloud-prod/workspace_staging_external_repo_metadata',
        'asset_key_parts': list([
          'aws',
          'cloud-prod',
          'workspace_staging_external_repo_metadata',
        ]),
        'description': 'Snowflake stages for AWS data, creates new stages for new assets, refreses existing stages.',
        'group_name': 'aws_stages',
        'id': 'dagster_open_platform.__repository__.["aws", "cloud-prod", "workspace_staging_external_repo_metadata"]',
        'kinds': list([
        ]),
        'metadata_entries': list([
        ]),
      }),
      dict({
        'asset_key': 'aws/cloud-prod/workspace_staging_jobs',
        'asset_key_parts': list([
          'aws',
          'cloud-prod',
          'workspace_staging_jobs',
        ]),
        'description': 'Snowflake stages for AWS data, creates new stages for new assets, refreses existing stages.',
        'group_name': 'aws_stages',
        'id': 'dagster_open_platform.__repository__.["aws", "cloud-prod", "workspace_staging_jobs"]',
        'kinds': list([
        ]),
        'metadata_entries': list([
        ]),
      }),
      dict({
        'asset_key': 'aws/cloud-prod/workspace_staging_metadata',
        'asset_key_parts': list([
          'aws',
          'cloud-prod',
          'workspace_staging_metadata',
        ]),
        'description': 'Snowflake stages for AWS data, creates new stages for new assets, refreses existing stages.',
        'group_name': 'aws_stages',
        'id': 'dagster_open_platform.__repository__.["aws", "cloud-prod", "workspace_staging_metadata"]',
        'kinds': list([
        ]),
        'metadata_entries': list([
        ]),
      }),
      dict({
        'asset_key': 'aws/cloud-prod/workspace_staging_partitions',
        'asset_key_parts': list([
          'aws',
          'cloud-prod',
          'workspace_staging_partitions',
        ]),
        'description': 'Snowflake stages for AWS data, creates new stages for new assets, refreses existing stages.',
        'group_name': 'aws_stages',
        'id': 'dagster_open_platform.__repository__.["aws", "cloud-prod", "workspace_staging_partitions"]',
        'kinds': list([
        ]),
        'metadata_entries': list([
        ]),
      }),
      dict({
        'asset_key': 'aws/cloud-prod/workspace_staging_pipelines',
        'asset_key_parts': list([
          'aws',
          'cloud-prod',
          'workspace_staging_pipelines',
        ]),
        'description': 'Snowflake stages for AWS data, creates new stages for new assets, refreses existing stages.',
        'group_name': 'aws_stages',
        'id': 'dagster_open_platform.__repository__.["aws", "cloud-prod", "workspace_staging_pipelines"]',
        'kinds': list([
        ]),
        'metadata_entries': list([
        ]),
      }),
    ]),
  })
# ---
# name: TestDynamicCommandExecution.test_command_execution[asset_success_nested_asset_json]
  dict({
    'asset_key': 'aws/cloud-prod/workspace_staging_pipelines',
    'asset_key_parts': list([
      'aws',
      'cloud-prod',
      'workspace_staging_pipelines',
    ]),
    'description': 'Snowflake stages for AWS data, creates new stages for new assets, refreses existing stages.',
    'group_name': 'aws_stages',
    'id': 'dagster_open_platform.__repository__.["aws", "cloud-prod", "workspace_staging_pipelines"]',
    'kinds': list([
    ]),
    'metadata_entries': list([
    ]),
  })
# ---
# name: TestDynamicCommandExecution.test_command_execution[asset_success_paginated_assets_json]
  dict({
    'cursor': '["__snowflake_query_metadata_self_serve_customer_attribution_check_all_partitions_job"]',
    'has_more': True,
    'items': list([
    ]),
  })
# ---
# name: TestDynamicCommandExecution.test_command_execution[asset_success_single_asset_json]
  dict({
    'asset_key': 'aws/cloud-prod/workspace_staging_pipelines',
    'asset_key_parts': list([
      'aws',
      'cloud-prod',
      'workspace_staging_pipelines',
    ]),
    'description': 'Snowflake stages for AWS data, creates new stages for new assets, refreses existing stages.',
    'group_name': 'aws_stages',
    'id': 'dagster_open_platform.__repository__.["aws", "cloud-prod", "workspace_staging_pipelines"]',
    'kinds': list([
    ]),
    'metadata_entries': list([
    ]),
  })
# ---
# name: TestDynamicCommandExecution.test_command_execution[deployment_empty_deployments_json]
  dict({
    'items': list([
      dict({
        'id': 6935,
        'name': 'uat',
        'type': 'PRODUCTION',
      }),
      dict({
        'id': 6936,
        'name': 'excessive',
        'type': 'PRODUCTION',
      }),
      dict({
        'id': 84915,
        'name': 'ben-test',
        'type': 'PRODUCTION',
      }),
      dict({
        'id': 37,
        'name': 'ecs',
        'type': 'PRODUCTION',
      }),
      dict({
        'id': 6934,
        'name': 'staging',
        'type': 'PRODUCTION',
      }),
      dict({
        'id': 80,
        'name': 'prod',
        'type': 'PRODUCTION',
      }),
    ]),
    'total': 6,
  })
# ---
# name: TestDynamicCommandExecution.test_command_execution[deployment_success_multiple_deployments_json]
  dict({
    'items': list([
      dict({
        'id': 6935,
        'name': 'uat',
        'type': 'PRODUCTION',
      }),
      dict({
        'id': 6936,
        'name': 'excessive',
        'type': 'PRODUCTION',
      }),
      dict({
        'id': 84915,
        'name': 'ben-test',
        'type': 'PRODUCTION',
      }),
      dict({
        'id': 37,
        'name': 'ecs',
        'type': 'PRODUCTION',
      }),
      dict({
        'id': 6934,
        'name': 'staging',
        'type': 'PRODUCTION',
      }),
      dict({
        'id': 80,
        'name': 'prod',
        'type': 'PRODUCTION',
      }),
    ]),
    'total': 6,
  })
# ---
# name: TestDynamicCommandExecution.test_command_execution[deployment_success_multiple_deployments_text_text]
  '''
  Name: uat
  ID: 6935
  Type: PRODUCTION
  
  Name: excessive
  ID: 6936
  Type: PRODUCTION
  
  Name: ben-test
  ID: 84915
  Type: PRODUCTION
  
  Name: ecs
  ID: 37
  Type: PRODUCTION
  
  Name: staging
  ID: 6934
  Type: PRODUCTION
  
  Name: prod
  ID: 80
  Type: PRODUCTION
  
  '''
# ---
